app/models/shipit/deploy_spec/npm_discovery.rb

```diff
@@ -79,10 +85,84 @@ def package_version
         JSON.parse(package_json.read)['version']
       end
 
+      def publish_config
+        JSON.parse(package_json.read)['publishConfig']
+      end
+
+      def publish_config_access
+        config = publish_config
+
+        # default to private deploy when we enforce a publishConfig
+        if enforce_publish_config?
+          return PRIVATE if config.blank?
+          config['access'] || PRIVATE
+        end
+
+        return PUBLIC if config.blank?
+        config['access'] || PUBLIC
+      end
+
+      def package_name
+        JSON.parse(package_json.read)['name']
+      end
+
+      def scoped_package?
+        package_name.start_with?('@shopify')
+      end
+
+      def enforce_publish_config?
+        enforce = Shipit.enforce_publish_config
+        return false if enforce.nil? || enforce.to_s == "0"
+        true
+      end
+
+      def valid_publish_config?
+        return true unless enforce_publish_config?
+        return false if Shipit.private_npm_registry.nil?
+        return false if publish_config.blank?
+        return true if publish_config_access == PUBLIC
+
+        valid_publish_config_access? && private_scoped_package?
+      end
+
+      def valid_publish_config_access?
+        VALID_ACCESS.include?(publish_config_access)
+      end
+
+      # ensure private packages are scoped
+      def private_scoped_package?
+        publish_config_access == PRIVATE && scoped_package?
+      end
+
+      def local_npmrc
+        file(".npmrc")
+      end
+
+      def npmrc_contents(registry)
+        "always-auth=true\n#{registry}"
+      end
+
+      def registry
+        prefix = scoped_package? ? "@shopify:registry" : "registry"
+        if publish_config_access == PUBLIC
+          return "#{prefix}=#{NPM_REGISTRY}"
+        end
+
+        "#{prefix}=#{Shipit.private_npm_registry}"
+      end
+
+      def generate_local_npmrc
+        contents = npmrc_contents(registry)
+        File.write(local_npmrc, contents) unless local_npmrc.exist?
```

Comment:
That seem wrong. A discovery module shouldn't write content on disk. And it's not executed at the proper time & place anyway so it won't be persisted.

In short, when a commit is merged we clone the repo, run the discovery modules, cache the result as JSON and delete the repo.


So if a file has to be written, it should be done by one of the commands executed.

Reply:
We spoke offline regarding this. 
This has been addressed with a shell script (https://github.com/Shopify/shipit-engine/pull/767/commits/be7ba76e950d71742a8785c0224e9a095485328f)